File: web-app/backend/app/sheets/portfolio.py
from app.sheets.client import get_sheet
from app.sheets.utils import to_float, to_float_or_none


def fetch_portfolio() -> list[dict]:
    """ポートフォリオシートから全保有銘柄を取得"""
    sheet = get_sheet("PORTFOLIO")
    records = sheet.get_all_records()
    result = []
    for r in records:
        # 空行スキップ
        if not r.get("銘柄コード"):
            continue

        acquired_price_foreign = to_float_or_none(r.get("取得単価（外貨）"))
        acquired_exchange_rate = to_float_or_none(r.get("取得時為替レート"))
        shares = to_float(r.get("保有株数", 0))

        # D列（取得単価（円））は数式 =E*F のため未評価時は外貨×為替でフォールバック
        acquired_price_jpy = to_float(r.get("取得単価（円）", 0))
        if (acquired_price_jpy == 0
                and acquired_price_foreign and acquired_exchange_rate):
            acquired_price_jpy = acquired_price_foreign * acquired_exchange_rate

        # H列（取得額合計）は数式 =D*G のため未評価時は単価×株数でフォールバック
        total_cost = to_float(r.get("取得額合計", 0))
        if total_cost == 0:
            total_cost = acquired_price_jpy * shares

        # 外国株フラグは Unicode コードポイント不一致や空白混入に備えて正規化
        flag_raw = str(r.get("外国株フラグ", "")).strip()
        is_foreign = flag_raw in ("○", "〇", "True", "true", "1")

        result.append({
            "code": r["銘柄コード"],
            "name": r["銘柄名"],
            "acquiredDate": str(r.get("取得日", "")),
            "acquiredPriceJpy": acquired_price_jpy,
            "acquiredPriceForeign": acquired_price_foreign,
            "acquiredExchangeRate": acquired_exchange_rate,
            "shares": shares,
            "totalCost": total_cost,
            "currency": str(r.get("通貨", "JPY")),
            "isForeign": is_foreign,
        })
    return result

File: portfolio-dashboard/server/src/routes/portfolio.ts
import { Hono } from "hono";
import { asc } from "drizzle-orm";
import { db } from "../db/index.js";
import { holdings, purchaseHistory } from "../db/schema.js";

const app = new Hono();

app.get("/", (c) => {
  const rows = db.select().from(holdings).all();

  // purchase_history を 1 クエリで全件取得し、code ごとに Map へグルーピング（N+1 禁止）
  const purchaseRows = db
    .select()
    .from(purchaseHistory)
    .orderBy(asc(purchaseHistory.seq))
    .all();

  const purchasesByCode = new Map<string, typeof purchaseRows>();
  for (const p of purchaseRows) {
    const list = purchasesByCode.get(p.code);
    if (list) {
      list.push(p);
    } else {
      purchasesByCode.set(p.code, [p]);
    }
  }

  const items = rows.map((r) => {
    let acquiredPriceJpy = r.acquiredPriceJpy;
    // acquiredPriceJpy が 0 で外貨フィールドがある場合はフォールバック計算
    if (
      acquiredPriceJpy === 0 &&
      r.acquiredPriceForeign != null &&
      r.acquiredExchangeRate != null
    ) {
      acquiredPriceJpy = r.acquiredPriceForeign * r.acquiredExchangeRate;
    }

    const totalCost = acquiredPriceJpy * r.shares;

    const purchases = (purchasesByCode.get(r.code) ?? []).map((p) => ({
      seq: p.seq,
      shares: p.shares,
      price: p.price,
      priceForeign: p.priceForeign ?? null,
      exchangeRate: p.exchangeRate ?? null,
      purchasedAt: p.purchasedAt,
    }));

    return {
      code: r.code,
      name: r.name,
      acquiredDate: r.acquiredDate ?? null,
      acquiredPriceJpy,
      acquiredPriceForeign: r.acquiredPriceForeign ?? null,
      acquiredExchangeRate: r.acquiredExchangeRate ?? null,
      shares: r.shares,
      totalCost,
      currency: r.currency,
      isForeign: r.isForeign,
      purchases,
    };
  });

  return c.json({ items });
});

export { app as portfolioRoute };
